docs/devElements.md

```diff
@@ -79,6 +79,7 @@ The `data` dictionary has the following possible keys (not all keys will be pres
 | `data["manual_grading"]`        | boolean | Whether the question is being rendered in the manual grading view.                                                                                  |
 | `data["panel"]`                 | string  | Which panel is being rendered (`question`, `submission`, or `answer`).                                                                              |
 | `data["extensions"]`            | dict    | A list of extensions that are available to be loaded by this element. For more information see the [element extensions](elementExtensions.md) page. |
+| `data["num_valid_submissions"]` | int     | The number of valid (not containing format errors) submissions by the student for the current variant.                                              |
```

Comment:
Is this correct that it is for the current variant? A lot of questions change variants after each attempt at them. In that case, I assume you still want to show hints after X number of tries, even if it is actually a different variant on each attempt, right?

Reply:
maybe instead of 'variant' it should be 'question instance'?

Reply:
hmm...I worry that this may be confusing for instructors. I can see instructors wanting to use this for questions where the student gets a new variant after every submission, but then being confused about why the hint isn't ever showing up after a few submissions. Unfortunately I don't really have a solution at the moment.

Reply:
I think revealing hints across variants has the potential to be useful, but I think this could also make it difficult to test for instructors without the ability to manually reset the number of attempts for a given user on the current assessment. I think functionality like that would be good to include, but I think requires some more care to make it user friendly.

Reply:
@SethPoulsen I think that we can simply document that this is how this element works: if you want to use hints, they'll only work within a certain variant. Can you think of a compelling pedagogical reason to give someone hints but only have those hints apply to different variants and not just multiple attempts to the same one?

Having question state that depends on previous variants instead of the current one is a precedent I really, really want to avoid setting. It just makes everything more complex, including implementation and conceptual models for students and faculty.

Reply:
I don't have statistics on that 😛 

I understand that's a thing that people _may_ do, but what's the pedagogical thinking for wanting hints _and_ multiple variants? Is the student experience or instructor's ability to measure mastery negatively impacted by having hints with only a single variant?

Reply:
Suppose the instructor wants to give a student 3 attempts at a question on a test, but they want it to be a different variant each time.

The hint is general enough that it applies to any of the variants of the same question, but they only want the student to see it on their last attempt. 

So, the instructor makes a hint with `show-after:3`, so that the students will get the hint for their last attempt. But to the instructor's surprise, the students _never_ see the hint because they only ever do 1 submission for each variant.

Reply:
Right, I'm not debating the fact that an instructor may want that; I'm trying to understand *why* they want multiple variants. Is it just personal preference? Or is there a specific reason they need that? Put differently, if I were to tell an instructor "sorry, hints can't span multiple variants", would they come back and tell me "I can't use a single variant because XYZ", or would they just happily switch to a single variant?

Reply:
I think people like using multiple variants in a situation like that because the student can be shown the correct answer and then learn from their mistake. If you do a single variant, it is harder for the student to learn from their mistake because you can't show them the correct answer.

Reply:
> If you do a single variant, it is harder for the student to learn from their mistake because you can't show them the correct answer.

This is really the core need this element is addressing and also why I'm skeptical of revealing hints across variants. There are questions where multiple variants are hard to generate or not appropriate, and the progressive hints is a good way of pushing students in the right direction without outright revealing the answer. If you're showing a new variant after each submission (possibly with a correct answer), this gives implicit feedback after each submission in a way that's not possible with single variant questions.

Reply:
I think that a common use case for hints would be with questions that generate new variants on every submission. I think instructors would naturally want to give a hint after (say) the third variant where the student got it wrong, to nudge them towards the correct method.

I'm not saying anything here about whether this is possible (or even should be possible) within an element --- I'm only saying what I think many instructors will want from a pedagogy point of view.

Reply:
> There are questions where multiple variants are hard to generate or not appropriate, and the progressive hints is a good way of pushing students in the right direction without outright revealing the answer.

Right, I'm totally with you here. I'm just saying that by the very fact that this element exists, people are going to try to use it for questions that give a new variant every submission (which is a totally reasonable thing to do), and then they're going to be frustrated when it doesn't work as expected. 

Maybe the way to make this shipp-able is just to be very clear about its limitations in the documentation. For example, change the summary of the element to include the fact that it won't work across variants, all the places in the documention that say 'the current question' should instead say 'the current variant', and the documentation for the `show-after` attribute should expclicitly state that `show-after` should only be non-zero with questions that have `singleVariant: true`.

Still, I'm sure we'll have feature requests from people who want it to work across variants.

Reply:
Yeah, I think this is going to be a support headache if it doesn't work as expected across variants.

Reply:
> Having question state that depends on previous variants instead of the current one is a precedent I really, really want to avoid setting. It just makes everything more complex, including implementation and conceptual models for students and faculty.

I get this (oh so much) but I think people's mental model is such that they will want this for hints. I think we should think hard about how to provide cross-variant functionality here, either by pulling hints functionality out into PL core, or by providing more data to elements so they can implement it.

Reply:
Going to go ahead and update the documentation as @SethPoulsen suggested. I think it's worth making a separate issue / discussion for how any type of cross-variant functionality might work. It should be easy enough to update this element to take advantage of that if or when those features get implemented.

Reply:
Resolving this discussion and created an issue for if/how we want to support this type of functionality in the future: https://github.com/PrairieLearn/PrairieLearn/issues/6465